src/AuditTrailModule/javasource/audittrail/log/CreateLogObject.java

```diff
@@ -296,115 +292,141 @@ private static IMendixObject createSingleLogLine(IMendixObject logObject, IMendi
 
-			return logLine;
+			return logLineList;
 		}
 		_logNode.trace("Skipping member: " + member.getName() + " because it has not changed.");
 		return null;
 	}
 
-	private static String getValueFromReference(IMendixIdentifier ID, IContext context) throws CoreException {
-		String value = "";
-
-		if (ID != null) {
-			// Get id of previous object
-			IMendixObject refObj = Core.retrieveId(context, ID);
-
-			if (refObj != null) {
-				// Get list of members with descriptive values
-				Collection<? extends IMendixObjectMember<?>> list = refObj.getMembers(context).values();
-
-				if (list.size() > 0) {
-					// loop the list
-					for (IMendixObjectMember<?> member : list) {
-						value += member.getName() + ": " + getValue(member, true, context) + "\n";
-					}
+	private static List<IMendixObject> createLogLinesForReferencedObject(IMendixIdentifier ID, IMendixIdentifier parentID,
+			IContext currentcontext, IContext sudocontext, TypeOfReferenceLog typeOfReference) throws CoreException {
+		if (ID == null) return Collections.emptyList();
+
+		// Get id of object
+		IMendixObject refObj = Core.retrieveId(currentcontext, ID);
+		List<IMendixObject> logLineList = new ArrayList<IMendixObject>();
+
+		if (refObj != null) {
+			// Get list of members with descriptive values
+			Collection<? extends IMendixObjectMember<?>> list = refObj.getMembers(currentcontext).values();
+
+			IMendixObject referenceLog = Core.instantiate(sudocontext, ReferenceLog.getType());
+			referenceLog.setValue(sudocontext, ReferenceLog.MemberNames.AttributeID.toString(), ID.toString());
+			referenceLog.setValue(sudocontext, ReferenceLog.MemberNames.Operation.toString(), typeOfReference.toString());
+			referenceLog.setValue(sudocontext, ReferenceLog.MemberNames.ReferenceLog_LogLine.toString(), parentID);
+			
+			logLineList.add(referenceLog);
+
+			if (list.size() > 0) {
+				// loop the list
+				for (IMendixObjectMember<?> member : list) {
+					IMendixObject referenceLogLine = Core.instantiate(sudocontext, ReferenceLogLine.getType());
+					referenceLogLine.setValue(sudocontext, ReferenceLogLine.MemberNames.Member.toString(), member.getName());
+					referenceLogLine.setValue(sudocontext, ReferenceLogLine.MemberNames.Value.toString(), getValue(member, true, currentcontext));
+					referenceLogLine.setValue(sudocontext, ReferenceLogLine.MemberNames.ReferenceLogLine_ReferenceLog.toString(), referenceLog.getId());
+					logLineList.add(referenceLogLine);
 				}
 			}
 		}
 
-		return value;
+		return logLineList;
 	}
+	
+	private static Comparator<IMendixIdentifier> IDCOMPARATOR = (IMendixIdentifier i1, IMendixIdentifier i2) -> (int)(i1.toLong() - i2.toLong());
 
+	// TODO The lists may be null. Check that.
 	@SuppressWarnings("unchecked")
-	private static IMendixObject createReferenceSetLogLine(IMendixObject logObject, IMendixObjectMember<?> member,
+	private static List<IMendixObject> createReferenceSetLogLine(IMendixObject logObject, IMendixObjectMember<?> member,

```

Comment:
```suggestion
private static List<IMendixObject> createReferenceSetLogLine(IMendixObject logObject, MendixObjectReferenceSet member,
			boolean isNew, IContext sudocontext, IContext currentcontext) throws CoreException {
```

If you apply this suggestion we can remove the `@SuppressWarnings("unchecked")` and you don't need to cast to `List<IMendixIdentifier>` at line 385 and 386.

Reply:
Done